Rainmaker library extension

sivar2311 opened the pull request:
## Description of Change
This PR extends the ESP Rainmaker functionality.
- Added definitions for various parameters and UI types.
- New overload of the `value` function with parameter type `const char*` to prevent a compiler warning with this type.
- New function `addValidStrList` added to Param class. This is required for using a Mode parameter.
- Added new RMakerCustomDeviceAirCooler example to demonstrate the use of Toggle, Mode and Range parameters.
  This is the Arduino version of the example from the chapter [Building Custom Devices / 3rd Party Integrations](https://rainmaker.espressif.com/docs/3rd-party.html).

## Tests scenarios
I have tested my Pull Request on Arduino-esp32 core v2.0.3 with esp32doit-devkit-v1 Board with the new provided RMakerCustomDeviceAirCooler example.

me-no-dev commented:
The changes you have done to the files in `tools/sdk/chip/...` can not be done in this repository. Files in that folder come from ESP-IDF or one of it's external components (esp-rainmaker in this case). You can either make a pull request to the [proper repository](https://github.com/espressif/esp-rainmaker) or add the definitions in the Arduino rainmaker headers.

sivar2311 commented:
Hi @me-no-dev!
Thanks for the review and feedback.
Yes, I was not sure about these files.

I don't know exactly how the workflow is from the ESP-IDF or Rainmaker repository to the arduino-esp32 repository.
To avoid code duplication and possible errors I think to do a pull request in the Rainmaker repository first.

Can this current pull request stay until the changes are applied to the Rainmaker / ESP-IDF repository, or is there a better way?

me-no-dev commented:
yup :) I've turned it into draft for now.

sivar2311 commented:
For tracking: PR in espressif/esp-rainmaker#126

sivar2311 commented:
Hi @me-no-dev The PR in espressif/esp-rainmaker#126 has been merged. 
What are the next steps to bring these changes to the arduino-esp32 repository?

me-no-dev commented:
Hi @sivar2311, thanks for letting me know. I have updated the lib-builder and generated the new libs. They are currently [here](https://github.com/espressif/arduino-esp32/pull/6861) and will soon be merged into master in preparation for 2.0.4 release. Please revert the changes to the headers in the "sdk" folder of this pull request as those will now come from the component itself through the lib builder. Also please squash the commits. I will run CI and merge as soon as we merge the IDF updates.

sivar2311 commented:
Hi @me-no-dev !
Thanks for the info. I will make the necessary changes as soon as possible.

Off-topic:
The whole construct from ESP-IDF to Arduino ESP32 is still very complex and obscure for me. Are there detailed descriptions or instructions that explain the relationships and processes? I would like to go deeper into the matter and contribute more.

me-no-dev commented:
This repository contains the ESP-IDF project and scripts that build ESP-IDF, collect and copy everything necessary to Arduino: https://github.com/espressif/esp32-arduino-lib-builder

sivar2311 commented:
@me-no-dev Thanks for the information about the esp32-arduino-lib-builder.
I have now reverted the changes to the header files from the "sdk" folder.

me-no-dev commented on libraries/RainMaker/src/RMakerParam.h:
```diff
@@ -47,5 +47,6 @@ class Param
         esp_err_t addUIType(const char *ui_type);
         esp_err_t addBounds(param_val_t min, param_val_t max, param_val_t step);
         esp_err_t updateAndReport(param_val_t val);
+        esp_err_t addValidStrList(const char** str_list);
```
your function declaration does not match the implementation. In the implementation you have two arguments: `const char **string_list, uint8_t count`. Probably you missed the second one.

me-no-dev commented:
@sivar2311 note that the target branch has changed to `idf-release/v4.4`, because that is where your changes are :)

sivar2311 commented:
@me-no-dev

> your function declaration does not match the implementation. In the implementation you have two arguments: `const char **string_list, uint8_t count`. Probably you missed the second one.

Sorry for the inconvenience, this is embarrassing. 
I saw the error message yesterday but I was already in bed. 

I already had the error during development and I had already fixed it - but probably only in my local platform directory (PlatformIO) which is used for the build process. Unfortunately I was not able to get PlattformIO to use my local copy of the Arduino-ESP32 repository for the build process.

I have fixed the bug in [#0cb567d](https://github.com/espressif/arduino-esp32/pull/6813/commits/0cb567deac2eec9aa74e33390bc9b979387d31d7)

> note that the target branch has changed to idf-release/v4.4, because that is where your changes are :)

Okay, that makes sense :) Do I need to change anything on my side?

me-no-dev commented:
Seems you have some other errors: https://github.com/espressif/arduino-esp32/runs/6875362072?check_suite_focus=true#step:5:164

> Okay, that makes sense :) Do I need to change anything on my side?
>

No, you should be ok as-is, just make sure you have updated your fork

sivar2311 commented:
The definition was missing for the ESP32C3 target. This is fixed now in [#de53f6f](https://github.com/espressif/arduino-esp32/pull/6813/commits/de53f6f1122ef26afc1192fb9eed7eb12282e92a)
Next time I will set up a better test suite that compiles and tests for all targets.

Thanks for your patience :)

me-no-dev commented:
@sivar2311 everything looks good now :) will merge before 2.0.4 release
